File: frontend/src/lib/catalog.js
import { badgeText } from '../i18n/ui';

export const getProductId = (product) => product?.id || product?._id || '';

const uniqueList = (items = []) => [
  ...new Set(items.map((item) => item?.toString().trim()).filter(Boolean)),
];

export const getProductCollectionSlugs = (product) =>
  uniqueList([
    product?.collectionSlug,
    ...(Array.isArray(product?.collectionSlugs)
      ? product.collectionSlugs
      : product?.collectionSlugs
      ? [product.collectionSlugs]
      : []),
  ]);

const normalizeVariant = (variant, index = 0, product = {}) => ({
  id: variant?.id || `${product.slug || getProductId(product) || 'variant'}-${index + 1}`,
  size: variant?.size || 'One Size',
  color: variant?.color || '',
  sku: variant?.sku || '',
  stock: Math.max(0, Number(variant?.stock || 0)),
  price:
    variant?.price === undefined || variant?.price === null || variant?.price === ''
      ? undefined
      : Number(variant.price),
  isActive: variant?.isActive ?? true,
});

const distributeStock = (totalStock, count, index) => {
  if (count <= 0) return 0;
  const base = Math.floor(totalStock / count);
  const remainder = totalStock % count;
  return base + (index < remainder ? 1 : 0);
};

const buildLegacyVariants = (product) => {
  const sizes = Array.isArray(product?.sizes) && product.sizes.length > 0 ? product.sizes : ['One Size'];
  const colors = Array.isArray(product?.colors) && product.colors.length > 0 ? product.colors : [''];
  const combinations = [];

  sizes.forEach((size) => {
    colors.forEach((color) => {
      combinations.push({ size, color });
    });
  });

  return combinations.map((combination, index) =>
    normalizeVariant(
      {
        id: `${product.slug || getProductId(product) || 'variant'}-${index + 1}`,
        size: combination.size,
        color: combination.color,
        sku: product?.sku ? `${product.sku}-${index + 1}` : '',
        stock: distributeStock(Math.max(0, Number(product?.stock || 0)), combinations.length, index),
      },
      index,
      product
    )
  );
};

export const normalizeProduct = (product) => {
  const normalizedImages = Array.isArray(product?.images) ? product.images : [];
  const collectionSlugs = getProductCollectionSlugs(product);
  const categories = uniqueList([
    product?.category,
    ...(Array.isArray(product?.categories) ? product.categories : product?.categories ? [product.categories] : []),
  ]);
  const normalizedVariants =
    Array.isArray(product?.variants) && product.variants.length > 0
      ? product.variants.map((variant, index) => normalizeVariant(variant, index, product))
      : buildLegacyVariants(product);
  const activeVariants = normalizedVariants.filter((variant) => variant.isActive !== false);

  return {
    ...product,
    id: getProductId(product),
    slug: product?.slug || getProductId(product),
    sku: product?.sku || '',
    status: product?.status || 'active',
    price: Number(product?.price || 0),
    compareAtPrice:
      product?.compareAtPrice === undefined || product?.compareAtPrice === null || product?.compareAtPrice === ''
        ? undefined
        : Number(product.compareAtPrice),
    stock: activeVariants.reduce((sum, variant) => sum + Math.max(0, Number(variant.stock || 0)), 0),
    trendingScore: Number(product?.trendingScore || 0),
    category: product?.category || categories[0] || '',
    categories,
    collectionSlug: collectionSlugs[0] || '',
    collectionSlugs,
    coverImage: product?.coverImage || normalizedImages[0] || '',
    images: normalizedImages,
    variants: normalizedVariants,
    sizes: [...new Set(activeVariants.map((variant) => variant.size).filter(Boolean))],
    colors: [...new Set(activeVariants.map((variant) => variant.color).filter(Boolean))],
    styleTags: Array.isArray(product?.styleTags) ? product.styleTags : [],
    seoTitle: product?.seoTitle || '',
    seoDescription: product?.seoDescription || '',
    isNew: Boolean(product?.isNew),
    isBestSeller: Boolean(product?.isBestSeller),
    isOnSale: Boolean(product?.isOnSale),
  };
};

export const normalizeProducts = (products = []) =>
  products.map(normalizeProduct).filter((product) => product.id);

export const titleize = (value = '') =>
  value
    .split('-')
    .filter(Boolean)
    .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
    .join(' ');

export const formatCurrency = (value) =>
  new Intl.NumberFormat('vi-VN', {
    style: 'currency',
    currency: 'VND',
    maximumFractionDigits: 0,
  }).format(Number(value || 0));

export const matchesCollection = (product, collection) => {
  if (!collection) return true;
  const collectionSlugs = getProductCollectionSlugs(product);
  if (collectionSlugs.length && collection?.slug) {
    return collectionSlugs.includes(collection.slug);
  }

  const source = [product.category, product.name, product.description]
    .filter(Boolean)
    .join(' ')
    .toLowerCase();

  return collection.featuredKeywords.some((keyword) => source.includes(keyword.toLowerCase()));
};

export const getProductBadges = (product, locale = 'vi') => {
  const badges = [];
  const labels = badgeText[locale] || badgeText.vi;

  if (product?.isNew) badges.push(labels.new);
  if (product?.isOnSale) badges.push(labels.sale);
  if (product?.isBestSeller) badges.push(labels.bestseller);
  if (Number(product?.stock || 0) > 0 && Number(product?.stock || 0) <= 8) badges.push(labels.lowStock);

  return badges;
};

export const isPublicProduct = (product) => product?.status !== 'draft';

export const getActiveVariants = (product) =>
  (product?.variants || []).filter((variant) => variant?.isActive !== false);

export const getAvailableVariants = (product) =>
  getActiveVariants(product).filter((variant) => Number(variant.stock || 0) > 0);

export const getVariantSizes = (product) =>
  [...new Set(getActiveVariants(product).map((variant) => variant.size).filter(Boolean))];

export const getVariantColors = (product, size) =>
  [
    ...new Set(
      getActiveVariants(product)
        .filter((variant) => !size || variant.size === size)
        .map((variant) => variant.color)
        .filter(Boolean)
    ),
  ];

export const getAvailableSizes = (product) =>
  [...new Set(getAvailableVariants(product).map((variant) => variant.size).filter(Boolean))];

export const getAvailableColors = (product, size) =>
  [
    ...new Set(
      getAvailableVariants(product)
        .filter((variant) => !size || variant.size === size)
        .map((variant) => variant.color)
        .filter(Boolean)
    ),
  ];

export const findVariant = (product, { size, color } = {}) =>
  getActiveVariants(product).find(
    (variant) =>
      (!size || variant.size === size) &&
      (!color || variant.color === color)
  ) || null;

export const findFirstAvailableVariant = (product) => getAvailableVariants(product)[0] || findVariant(product);

export const isVariantAvailable = (product, { size, color } = {}) =>
  Number(findVariant(product, { size, color })?.stock || 0) > 0;

File: backend/catalog.py
from typing import Any, Dict, Iterable, List

from bson import ObjectId


def is_object_id(value: str) -> bool:
    try:
        ObjectId(value)
        return True
    except Exception:
        return False


def slugify(value: str) -> str:
    return (
        str(value or "")
        .strip()
        .lower()
        .replace("&", " and ")
        .replace("/", " ")
        .replace("_", " ")
        .replace("  ", " ")
        .replace(" ", "-")
    )


def _clean_list(items: Iterable[Any]) -> List[str]:
    result = []
    seen = set()
    for item in items:
        value = str(item or "").strip()
        if not value or value in seen:
            continue
        seen.add(value)
        result.append(value)
    return result


def normalize_collection_slugs(product_doc: Dict[str, Any]) -> List[str]:
    collection_slugs = product_doc.get("collectionSlugs") or []
    if isinstance(collection_slugs, str):
        collection_slugs = [collection_slugs]
    return _clean_list([product_doc.get("collectionSlug"), *collection_slugs])


def _normalize_variants(product_doc: Dict[str, Any]) -> List[Dict[str, Any]]:
    variants = product_doc.get("variants") or []
    normalized = []

    for index, variant in enumerate(variants):
        normalized.append(
            {
                "id": variant.get("id") or f"{product_doc.get('slug') or product_doc.get('_id')}-variant-{index + 1}",
                "size": variant.get("size") or "One Size",
                "color": variant.get("color") or "",
                "sku": variant.get("sku") or "",
                "stock": max(0, int(variant.get("stock", 0) or 0)),
                "price": variant.get("price"),
                "isActive": variant.get("isActive", True),
            }
        )

    return normalized


def serialize_product(product_doc: Dict[str, Any]) -> Dict[str, Any]:
    doc = dict(product_doc)
    collection_slugs = normalize_collection_slugs(doc)
    fallback_collection_slug = slugify(doc.get("category") or "uncategorized")
    categories = _clean_list(doc.get("categories") or [doc.get("category")])
    variants = _normalize_variants(doc)
    active_variants = [variant for variant in variants if variant.get("isActive", True)]

    stock = sum(int(variant.get("stock", 0) or 0) for variant in active_variants)
    sizes = _clean_list(variant.get("size") for variant in active_variants if variant.get("size"))
    colors = _clean_list(variant.get("color") for variant in active_variants if variant.get("color"))

    serialized = {
        **doc,
        "id": str(doc.get("_id")),
        "slug": doc.get("slug") or str(doc.get("_id")),
        "sku": doc.get("sku") or "",
        "status": doc.get("status") or "active",
        "collectionSlug": collection_slugs[0] if collection_slugs else fallback_collection_slug,
        "collectionSlugs": collection_slugs,
        "compareAtPrice": doc.get("compareAtPrice"),
        "coverImage": doc.get("coverImage") or "",
        "images": doc.get("images") or [],
        "variants": variants,
        "styleTags": doc.get("styleTags") or [],
        "material": doc.get("material") or "",
        "fitNotes": doc.get("fitNotes") or "",
        "seoTitle": doc.get("seoTitle") or "",
        "seoDescription": doc.get("seoDescription") or "",
        "isNew": bool(doc.get("isNew", False)),
        "isBestSeller": bool(doc.get("isBestSeller", False)),
        "isOnSale": bool(doc.get("isOnSale", False)),
        "trendingScore": int(doc.get("trendingScore", 0) or 0),
        "category": doc.get("category") or (categories[0] if categories else ""),
        "categories": categories,
        "stock": stock if variants else max(0, int(doc.get("stock", 0) or 0)),
        "sizes": sizes if variants else _clean_list(doc.get("sizes") or []),
        "colors": colors if variants else _clean_list(doc.get("colors") or []),
    }

    serialized.pop("_id", None)
    return serialized


def serialize_collection(collection_doc: Dict[str, Any]) -> Dict[str, Any]:
    doc = dict(collection_doc)
    serialized = {
        **doc,
        "id": str(doc.get("_id")),
        "slug": doc.get("slug") or "",
        "title": doc.get("title") or "",
        "description": doc.get("description") or "",
        "image": doc.get("image") or "",
        "featuredKeywords": doc.get("featuredKeywords") or [],
        "seoHeading": doc.get("seoHeading") or "",
        "seoBody": doc.get("seoBody") or "",
        "sortPriority": int(doc.get("sortPriority", 99) or 99),
        "isActive": bool(doc.get("isActive", True)),
    }
    serialized.pop("_id", None)
    return serialized
